working/macros/feature-specification.md

```diff
@@ -550,6 +550,116 @@ This does have two interesting and possibly unexpected consequences:
 
 **TODO**: Define the API for adding metadata to existing declarations (#1931).
 
+## Aspect Macros
+
+**NOTE**: This feature is a work in progress, and has many TODOs.
+
+An AspectMacro is used by another macro, to generate shared helper code which
+will not be duplicated across every library where that helper was needed.
+
+Concretely, let's consider the following example:
+
+```dart
+@fromJsonExtension
+class A {}
+
+@fromJsonExtension
+class B {
+  final A a;
+
+  B(this.a);
+}
+
+// Generated extension from the macro on A
+extension AFromJson on Map<String, Object?> {
+  A toA();
+}
+
+// Generated extension from the macro on B
+extension BFromJson on Map<String, Object?> {
+  A toA(); // This also needs to generate a helper for converting `A`!
+
+  B toB();
+}
+```
+
+Both of these macro applications need to generate the same shared code for
+converting a Map into an instance of `A` (since `B` has a field of that type),
+but they can't actually _see_ what each other has generated.
```

Comment:
There might be another argument that requires this aspect macro approach, but specifically this example does not look convincing to me.

I was triggered by `A special library "next" to each library` statement below. So, we kind of pretend that something was generated while running phases 1-3, but actually we don't want to... why?

If `fromJsonExtension` runs on `B` and find that it needs something from `A`, then this something should have been generated while running on `A`. And `B` just knows that it was generated and uses it. If there is no required method (or whatever) for `A`, the macro reports an error.

Reply:
> If `fromJsonExtension` runs on `B` and find that it needs something from `A`, then this something should have been generated while running on `A`. And `B` just knows that it was generated and uses it. If there is no required method (or whatever) for `A`, the macro reports an error.

The general use case is where you might want to generate something "on the side" of another type, without having to add it explicitly to the library where that type is defined. This could be serialization logic or anything else.

This pattern has several advantages:

- Users who do not need this functionality do not pay the cost for it.
- You don't have to try and get your  dependencies to support said functionality. You can add it yourself (as long as you can add it via new top level declarations of some kind).
- You can get all of this without causing code bloat/duplication when the same helper code is needed by multiple macros (using this feature at least).

For instance consider the use case of `mockito`. Classes generally should not all generate mocks and ship them as a part of their library. Instead, it is better if tests ask for the specific mocks they need, and that is how mockito works today. You can generate transitive mocks for a given type with an annotation like `@GenerateMocks([Cat])` on any top level declaration in your test file. This has to generate N classes (one for each user defined type that appears in the API of Cat basically). You can't actually do this at all today, you would have to manually specify all the transitive mocks to create in the annotation.

> I was triggered by `A special library "next" to each library` statement below. So, we kind of pretend that something was generated while running phases 1-3, but actually we don't want to... why?

We don't actually run it as a part of those phases in order to allow for multiple macros  to run the same "Aspect Macro", but only actually run it once. Also, these aspect macros have full access to introspect on the program (except the result of other aspect macros). Basically, it's a 4th phase which works very differently from the other phases, but the output of which is not introspectable beyond just getting an identifier. You can iteratively ask for more Aspect Macros to be applied in this phase as well, which is another key difference. This is safe to do because of the lack of introspection capability.

The particulars of where this code would end up in generally is tricky, I think conceptually the code is "next" to each library (this feature is inspired heavily by [bazel aspects](https://docs.bazel.build/versions/main/skylark/aspects.html)). But the code obviously can't actually live there. I also don't think we want to jam all of these aspects into a single top level library.